ocaml/xapi/xapi_cluster.ml

```diff
@@ -194,17 +194,14 @@ let pool_destroy ~__context ~self =
   Helpers.call_api_functions ~__context (fun rpc session_id ->
       Client.Client.Cluster.destroy ~rpc ~session_id ~self)
 
-let pool_resync ~__context ~self =
-  (* First ensure that the Cluster_hosts that are enabled in xapi's DB are really enabled *)
-  Db.Cluster_host.get_all ~__context |> List.iter (fun cluster_host ->
-      if Db.Cluster_host.get_enabled ~__context ~self:cluster_host then
-        Helpers.call_api_functions ~__context (fun rpc session_id ->
-            (* We rely on the fact that Cluster_host.enable unconditionally
-               invokes the low-level enable operations and is idempotent. *)
-            Client.Client.Cluster_host.enable ~rpc ~session_id ~self:cluster_host)
-    );
-  (* Then create the missing Cluster_hosts *)
-  let pool_auto_join = Db.Cluster.get_pool_auto_join ~__context ~self in
-  if pool_auto_join then begin
-      Db.Host.get_all ~__context |> List.iter (fun host -> Xapi_cluster_host.create_as_necessary ~__context ~host)
-  end
+let pool_resync ~__context ~(self : API.ref_Cluster) =
+  List.iter
+    (fun host -> log_and_ignore_exn
+        (fun () ->
+           Xapi_cluster_host.resync_host ~__context ~host;
+           if is_clustering_disabled_on_host ~__context host
+           then raise Api_errors.(Server_error (no_compatible_cluster_host, [Ref.string_of host]))
+            (* If host.clustering_enabled then resync_host should successfully
+               find or create a matching cluster_host which is also enabled *)
+        )
+    ) (Xapi_pool_helpers.get_master_slaves_list ~__context)
```

Comment:
I think error reporting could be improved here, but that can be done in a separate PR that goes through each clustering API and thinks about how error handling/reporting should look like for it.
We should probably have a Xapi_cluster_host.health_check and Xapi_cluster.health_check API that does some basic consistency checks (for cluster host it would check whether clustering is enabled and call a cluster_interface API diagnostics command and check if all good there), similarly the Xapi_cluster.health_check would check the health of all cluster hosts via their individual health check API, and also check whether we have cluster hosts for all hosts if auto join is enabled. Please add an entry about this on your checklist of GFS2 cleanups.